### PR TITLE
Api platform payment

### PR DESCRIPTION
**Pull request by SirDomin**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets |
| License         | MIT

Introduced payment methods to api-platform, tests are WIP, base logic presented. Once we complete payment, PaymentConfigurationProvider.php will return payment options based on selected payment method, example: https://github.com/Sylius/PayPalPlugin/pull/170  
With this data we can complete payment and go through PayPal website with authorization.

example response from api-platform:
image

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        $this->orderRepository = $orderRepository;
+        $this->paymentConfigurationProvider = $paymentConfigurationProvider;
+    }
+
+    public function __invoke(Request $request): Response
+    {
+        /** @var OrderInterface $order */
+        $order = $this->orderRepository->find($request->get('id'));
+
+        return $this->paymentConfigurationProvider->provide($order->getLastPayment());
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:43]: what if order has more than one payment?

File: src/Sylius/Bundle/ApiBundle/Payment/ApiPaymentMethodInterface.php [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<?php
+
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Payment/ApiPaymentMethodInterface.php:2]: missing licence block

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/OrderPayment.xml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ For the full copyright and license information, please view the LICENSE
+ file that was distributed with this source code.
+
+-->
+
+<resources xmlns="https://api-platform.com/schema/metadata"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="https://api-platform.com/schema/metadata https://api-platform.com/schema/metadata/metadata-2.0.xsd"
+>
+    <resource class="%sylius.model.order.class%" shortName="OrderPayment">
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/OrderPayment.xml:18]: sth is wrong with this config 🗡️  it is named as `OrderPayment`, class is an `order` short name is `OrderPayment`, and this `GET` function returns `paymentConfiguration` :D
Definitely, we need to disuse it on audio because I am not sure what do you want to achieve here :D

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/api_payment_methods.xml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+-->
+
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd"
+>
+    <services>
+        <service id="Sylius\Bundle\ApiBundle\Provider\PaymentConfigurationProvider">
+            <argument type="tagged_iterator" tag="api.payment" />
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/api_payment_methods.xml:20]: Can you explain why did you add here this tag? and how is it work?

**Review by AdamKasp — CHANGES_REQUESTED** [review]

File: src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php (hunk 1 above)
Comment (SirDomin) [reply src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:43]: it gets last added payment

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/api_payment_methods.xml (hunk 4 above)
Comment (SirDomin) [reply src/Sylius/Bundle/ApiBundle/Resources/config/services/api_payment_methods.xml:20]: Alright, once we create Class with interface ApiPaymentMethodInterface we can tag this service in our plugin or payment extension and it will load it automatically to PaymentConfigurationProvider as it takes all services tagged with api.payment. When this interface is in core there is another thing to do and auto tag every single class which implements ApiPaymentMethodInterface, this way there wont be need for tagging every service, all you need to do is provide class with given interface and our provider will take care of it ;) (generic solution i guess xD)

Comment (SirDomin) [reply src/Sylius/Bundle/ApiBundle/Resources/config/services/api_payment_methods.xml:20]: but for now tagging every service works enough, ill remember about refactor before we implement any other payments to api

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/OrderPayment.xml (hunk 3 above)
Comment (SirDomin) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/OrderPayment.xml:18]: yes, thats true. I wasnt sure tho how to handle it proper way so i left it with my first concept (working one) as there's some problems like unability to get payment resource as guest :/

File: src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace Sylius\Bundle\ApiBundle\Controller\Payment;
+
+use Sylius\Bundle\ApiBundle\Provider\PaymentConfigurationProvider;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Repository\OrderRepositoryInterface;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Request;
+
+final class GetPaymentConfiguration
```
Comment (Tomanhez) [inline src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:22]: Add `@experimental` blocks, here and in other missing places

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/OrderPayment.xml (hunk 3 above)
Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/OrderPayment.xml:18]: Maybe you should think about rout like `/new-api/shop/orders/{id}/payments/{paymentId}/configuration` ? We have similar functionality in `/new-api/shop/orders/{id}/payments/{paymentId}` and guest can get payment. Its resolved in `Sylius\Bundle\ApiBundle\Command\Checkout\ChoosePaymentMethod` and his handler. If you would take this solution you can remove this file and custom controller next all this logic can be handle in command and command handler

File: src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        $this->orderRepository = $orderRepository;
+        $this->paymentConfigurationProvider = $paymentConfigurationProvider;
+    }
+
+    public function __invoke(Request $request): JsonResponse
+    {
+        /** @var OrderInterface $order */
+        $order = $this->orderRepository->find($request->get('id'));
+
+        return new JsonResponse($this->paymentConfigurationProvider->provide($order->getLastPayment()));
```
Comment (Tomanhez) [inline src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:44]: Is good idea to get last payment? Maybe we should be able to add needed payment `id` in the path? I continue idea from my comment below with using `/new-api/shop/orders/{id}/payments/{paymentId}/configuration`. What is your opinion?

File: src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        PaymentConfigurationProvider $paymentConfigurationProvider
+    ) {
+        $this->orderRepository = $orderRepository;
+        $this->paymentConfigurationProvider = $paymentConfigurationProvider;
+    }
+
+    public function __invoke(Request $request): JsonResponse
+    {
+        /** @var OrderInterface $order */
+        $order = $this->orderRepository->find($request->get('id'));
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:42]: We should look for order by token or numer

File: src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php (hunk 6 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:44]: If we want to go with the last payment, we should use the last new payment. But your proposed path seems like a better approach

File: src/Sylius/Bundle/ApiBundle/Payment/ApiPaymentMethodInterface.php [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Payment;
+
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\Model\PaymentMethodInterface;
+
+interface ApiPaymentMethodInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Payment/ApiPaymentMethodInterface.php:19]: PaymentCongifurationProvider?

File: src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+{
+    /** @var OrderRepositoryInterface */
+    private $orderRepository;
+
+    /** @var PaymentConfigurationProvider */
+    private $paymentConfigurationProvider;
+
+    public function __construct(
+        OrderRepositoryInterface $orderRepository,
+        PaymentConfigurationProvider $paymentConfigurationProvider
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:33]: Payment configuration is an implementation of the class, not a contract. We should introduce an interface for it.

File: src/Sylius/Bundle/ApiBundle/Provider/PaymentConfigurationProvider.php [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\Model\PaymentMethodInterface;
+
+/** @experimental */
+final class PaymentConfigurationProvider
+{
+    /** @var iterable */
+    private $apiPayments;
+
+    public function __construct(iterable $apiPayments)
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Provider/PaymentConfigurationProvider.php:25]: Can we check if all providers have proper interface?

Comment (lchrusciel) [reply src/Sylius/Bundle/ApiBundle/Provider/PaymentConfigurationProvider.php:25]: The variable should have a more meaningful name as well

File: src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php (hunk 9 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:33]: Maybe we can use the interface, that you have introduced?

File: src/Sylius/Bundle/ApiBundle/Provider/PaymentConfigurationProvider.php [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Provider;
+
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\Model\PaymentMethodInterface;
+
+/** @experimental */
+final class PaymentConfigurationProvider
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Provider/PaymentConfigurationProvider.php:20]: Composit as a prefix

File: src/Sylius/Bundle/ApiBundle/Payment/ApiPaymentMethodInterface.php (hunk 8 above)
Comment (SirDomin) [reply src/Sylius/Bundle/ApiBundle/Payment/ApiPaymentMethodInterface.php:19]: this is the interface we should use for any payment implementation

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml [hunk 12]
```diff
@@ -221,6 +221,33 @@
                 </attribute>
             </itemOperation>
 
+            <itemOperation name="get_configuration">
+                <attribute name="method">GET</attribute>
+                <attribute name="controller">sylius.api.get_configuration_action</attribute>
+                <attribute name="path">/shop/orders/{id}/payments/{paymentId}/configuration</attribute>
```
Comment (Tomanhez) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml:227]:
```suggestion
                <attribute name="path">/shop/orders/{id}/payments/{paymentId}/configuration</attribute>
```
Is order `id` used anywhere ?

Comment (SirDomin) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml:227]: i guess it is, this was we secure endpoint for people with order token (we dont have it in payments)

File: src/Sylius/Bundle/ApiBundle/Payment/CompositeApiPaymentMethodInterface.php [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Payment;
+
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\Model\PaymentMethodInterface;
+
+/** @experimental */
+interface CompositeApiPaymentMethodInterface
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Payment/CompositeApiPaymentMethodInterface.php:20]: The `Composite` prefix should be in provider service, that iterates over a collection of payment handlers

```suggestion
interface ApiPaymentMethodHandlerInterface
```

File: src/Sylius/Bundle/ApiBundle/Provider/PaymentConfigurationProvider.php [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Provider;
+
+use Sylius\Bundle\ApiBundle\Payment\CompositeApiPaymentMethodInterface;
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\Model\PaymentMethodInterface;
+
+/** @experimental */
+final class PaymentConfigurationProvider implements  PaymentConfigurationProviderInterface
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Provider/PaymentConfigurationProvider.php:21]:
```suggestion
final class CompositePaymentConfigurationProvider implements  PaymentConfigurationProviderInterface
```

File: src/Sylius/Bundle/ApiBundle/Provider/PaymentConfigurationProvider.php [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        $this->apiPaymentHandlers = $apiPaymentHandlers;
+    }
+
+    public function provide(PaymentInterface $payment): array
+    {
+        /** @var PaymentMethodInterface $paymentMethod */
+        $paymentMethod = $payment->getMethod();
+
+        foreach ($this->apiPaymentHandlers as $apiPayment) {
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Provider/PaymentConfigurationProvider.php:36]:
```suggestion
        foreach ($this->apiPaymentMethodHandlers as $apiPaymentMethodHandler) {
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/api_payment_methods.xml (hunk 4 above)
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/api_payment_methods.xml:20]:
```suggestion
            <argument type="tagged_iterator" tag="sylius.api.payment_method_handler" />
```

File: src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    public function __invoke(Request $request): JsonResponse
+    {
+        /** @var PaymentInterface|null $payment */
+        $payment = $this->paymentRepository->find($request->get('paymentId'));
+
+        Assert::notNull($payment);
+
+        return new JsonResponse($this->paymentConfigurationProvider->provide($payment));
```
Comment (Tomanhez) [inline src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:47]:
```suggestion
        $payment = $this->paymentRepository->find($request->get('paymentId'));

        Assert::notNull($payment);

        return new JsonResponse($this->paymentConfigurationProvider->provide($payment));
```

This endpoint is opened for all users, login and not. We have access to payment `id` and order `id` and we should check next steps:
1) get user form userContext 
2) check login status
3) get order and assert order owner
4) if user is order owner next we should check if payment is on this order 
If all above conditions are met we can response configuration
Similar partially configuration is here: https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/ApiBundle/DataProvider/OrderCollectionDataProvider.php

File: src/Sylius/Bundle/ApiBundle/Provider/PaymentConfigurationProvider.php (hunk 14 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Provider/PaymentConfigurationProvider.php:21]:
```suggestion
final class PaymentConfigurationProvider implements PaymentConfigurationProviderInterface
```

File: src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php (hunk 16 above)
Comment (SirDomin) [reply src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php]: i've created order as a user and i couldnt access it without jwt token, we check if payment belongs to order now.

File: src/Sylius/Bundle/ApiBundle/Payment/CompositeApiPaymentMethodInterface.php (hunk 13 above)
[the remark this answers is not in the document]
Comment (SirDomin) [reply src/Sylius/Bundle/ApiBundle/Payment/CompositeApiPaymentMethodInterface.php]: i changed it to ApiPaymentMethodHandlerInterface

File: src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+        PaymentConfigurationProviderInterface $paymentConfigurationProvider
+    ) {
+        $this->paymentRepository = $paymentRepository;
+        $this->paymentConfigurationProvider = $paymentConfigurationProvider;
+    }
+
+    public function __invoke(Request $request): JsonResponse
+    {
+        /** @var PaymentInterface|null $payment */
+        $payment = $this->paymentRepository->find($request->get('paymentId'));
```
[the remark this answers is not in the document]
Comment (SirDomin) [reply src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:50]: it doesnt work when i change it to your proposition :O

**Review by arti0090 — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+
+        /** @var OrderInterface|null $order */
+        $order = $this->orderRepository->findOneByTokenValue($request->get('id'));
+
+        Assert::notNull($order);
+        Assert::notNull($payment);
```
Comment (arti0090) [inline src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:56]: I would move `Assert::notNull($payment);` just under `$payment = $this->paymentRepository->find($request->get('paymentId'));` so in case $payment is null we want call DB second time. In other way grouped asserts looks nicer so it is up to you 😄

Comment (SirDomin) [reply src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:56]: you're right, but its cached and i dont think its a big deal tho. Code looks better this way i'd say

Comment (GSadee) [reply src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:56]: I would be even for only one call to DB to get payment for given id and order token value. Or you can get order firstly, and then get the payment for id and order, in this case, there is no need for custom repository method, but there will be 2 calls to DB.

File: src/Sylius/Bundle/ApiBundle/Payment/ApiPaymentMethodHandlerInterface.php [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Payment;
+
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\Model\PaymentMethodInterface;
+
+/** @experimental */
+interface ApiPaymentMethodHandlerInterface
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Payment/ApiPaymentMethodHandlerInterface.php:20]: I wonder if the better name would not be:
```suggestion
interface ApiPaymentConfigurationProviderInterface
```
or sth like that, but as it is an experimental feature, it could be refactored later

**Review by GSadee — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Payment/ApiPaymentMethodHandlerInterface.php (hunk 19 above)
Comment (AdamKasp) [reply src/Sylius/Bundle/ApiBundle/Payment/ApiPaymentMethodHandlerInterface.php:20]: IMHO we shouldn't leave not perfect names because it is `the experimental feature`, we can leave it as it is if this name is good enough to merge :D

**Review by AdamKasp — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php (hunk 17 above)
[the remark this answers is not in the document]
Comment (lchrusciel) [reply src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:50]: ah, because `paymentId` is in the URL
```suggestion
        $payment = $this->paymentRepository->find($request->attributes->get('paymentId'));
```

File: src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php (hunk 18 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:56]: 1 call to the database based on order token and paymentId FTW

File: src/Sylius/Bundle/ApiBundle/Payment/ApiPaymentMethodHandlerInterface.php (hunk 19 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/ApiBundle/Payment/ApiPaymentMethodHandlerInterface.php:20]: The service is already in ApiBundle and therefore, API prefix for this interface seems for me redundant. Also, these providers do not return anything API based. It provides an array of configurations, which will be used in API context later. Service itself is not coupled to API itself
```suggestion
interface PaymentConfigurationProviderInterface
```

File: src/Sylius/Bundle/ApiBundle/Provider/CompositePaymentConfigurationProvider.php [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+        $this->apiPaymentMethodHandlers = $apiPaymentMethodHandlers;
+    }
+
+    public function provide(PaymentInterface $payment): array
+    {
+        /** @var PaymentMethodInterface $paymentMethod */
+        $paymentMethod = $payment->getMethod();
+
+        foreach ($this->apiPaymentMethodHandlers as $apiPaymentMethodHandler) {
+            if ($apiPaymentMethodHandler->supports($paymentMethod)) {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Provider/CompositePaymentConfigurationProvider.php:37]: Maybe instead of the support method, we can trigger UnsuppportedPaymentMethodException(a custom one) or use something more like delegator patter. The second approach would mean, that we are injecting an array of services but each item is identified by some key (like payment method config). In both cases, we can drop `supports` method and composite itself can implement the same interface as end providers itself.

File: src/Sylius/Bundle/ApiBundle/Provider/CompositePaymentConfigurationProvider.php [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Provider;
+
+use Sylius\Bundle\ApiBundle\Payment\ApiPaymentMethodHandlerInterface;
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\Model\PaymentMethodInterface;
+
+/** @experimental */
+final class CompositePaymentConfigurationProvider implements PaymentConfigurationProviderInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Provider/CompositePaymentConfigurationProvider.php:21]:
```suggestion
final class CompositePaymentConfigurationProvider implements CompositPaymentConfigurationProviderInterface
```

The current interface seems to be a better fit here: https://github.com/Sylius/Sylius/pull/12107/files#diff-891e0e19d6fdd944d01eca4ac4e136f9c24d86cb111a0e4a655d8a2465b53e59R20. It could stay as it is if we would drop support method (mentioned in previous comment)

**Comment by SirDomin** [comment]
Introduced repository method to find payment based on order token value 💯

File: src/Sylius/Bundle/ApiBundle/Provider/CompositePaymentConfigurationProvider.php (hunk 20 above)
Comment (SirDomin) [reply src/Sylius/Bundle/ApiBundle/Provider/CompositePaymentConfigurationProvider.php:37]: i would say it all defends itslef, we have handlers. If they support payment method it means they can provide data based on payment method. its clearer to read if supports then do something rather that inserting it into try catch block.

**Review by arti0090 — APPROVED** [review]

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, @SirDomin! :1st_place_medal:

File: src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+        CompositePaymentConfigurationProviderInterface $compositePaymentConfigurationProvider
+    ) {
+        $this->paymentRepository = $paymentRepository;
+        $this->compositePaymentConfigurationProvider = $compositePaymentConfigurationProvider;
+    }
+
+    public function __invoke(Request $request): JsonResponse
+    {
+        /** @var PaymentInterface|null $payment */
+        $payment = $this->paymentRepository->findOneByOrderToken($request->get('paymentId'), $request->get('id'));
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Controller/Payment/GetPaymentConfiguration.php:45]:
```suggestion
        $payment = $this->paymentRepository->findOneByIdAndOrderTokenValue($request->get('paymentId'), $request->get('id'));
```

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/PaymentRepository.php [hunk 23]
```diff
@@ -41,6 +41,19 @@ public function findOneByOrderId($paymentId, $orderId): ?PaymentInterface
         ;
     }
 
+    public function findOneByOrderToken(string $paymentId, string $orderToken): ?PaymentInterface
+    {
+        return $this->createQueryBuilder('p')
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/PaymentRepository.php:46]: As this undocumented best practice, we are using `o` for default query builder
```suggestion
        return $this->createQueryBuilder('o')
```
